drivers/flash/nrf_qspi_nor.c

```diff
@@ -491,15 +491,36 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
 		return -EINVAL;
 	}
 
-	/* read size must be non-zero multiple of 4 bytes */
-	if (((size % 4U) != 0) || (size == 0)) {
-		return -EINVAL;
-	}
 	/* address must be 4-byte aligned */
 	if ((addr % 4U) != 0) {
```

Comment:
So what you are suggesting have been solved here https://github.com/NordicSemiconductor/nrfx/blob/59ad34410bf79b9b50b86c8befb9d71024f04c0b/drivers/src/nrfx_nvmc.c#L385-L446

So the question is should this be solved in the nrfx driver in the future or should it be done here?

Reply:
I don't know what nrfx does, or whether the QSPI wrapper in it should try to work around hardware alignment limitations as it apparently did for NVMC.

The case I was concerned about was reading a single byte at, say, address 0x03, which was not supported by zephyrproject-rtos/zephyr#25292 which I'd originally hoped to get into Zephyr 2.3.0 to fix the known mcuboot problem.   But it didn't get reviews, and even though it fixes mcuboot it's not as general as it could be.

Your approach is more complete, but it's not going into upstream 2.3.0 either, and if you think it's temporary I don't know what's going on or what the schedule is.  If the intent is to remove the alignment requirements at the nrfx level, why not start there?

Reply:
I don't know if the intent is to remove the alignment requirement at nrfx. I'll ask around :) 

We need to be able to write data from flash to fix MCUboot problem completely. Since writing the BOOT_TEST word uses data in flash.

Reply:
But I could add that on what you already have upstream.

Reply:
Could you comment on that issue explaining what you think is missing (i.e. where a sub-word write could happen)?  I'm pretty sure I did test image swapping with image-1 on the external flash and it worked.